net/libfab/libfab.c

```diff
@@ -2494,22 +2575,120 @@ static int libfab_txep_init(struct m0_fab__active_ep *aep,
+	int                     i;
+
+	M0_PRE(old_ctx != NULL && old_fid != NULL && old_size < new_size);
+
+	M0_ALLOC_ARR(new_ctx, new_size);
+	M0_ALLOC_ARR(new_fid, new_size);
+	if (new_ctx == NULL || new_fid == NULL) {
+		m0_free(new_ctx);
+		m0_free(new_fid);
+		return M0_ERR(-ENOMEM);
+	}
+
+	/* Copy fids from old array to new array. */
+	for (i = 0; i < old_size; i++) {
+		new_ctx[i] = old_ctx[i];
+		new_fid[i] = old_fid[i];
+	}
+	tmfid->ftf_ctx      = new_ctx;
+	tmfid->ftf_head     = new_fid;
+	tmfid->ftf_arr_size = new_size;
+
+	M0_LOG(M0_DEBUG,"old={fid=%p ctx=%p size=%d} new={fid=%p ctx=%p size=%d}",
+	       old_fid, old_ctx, old_size, new_fid, new_ctx, new_size);
+
+	/* Free old array */
+	m0_free(old_ctx);
+	m0_free(old_fid);
+
+	return M0_RC(0);
+}
+
+/**
+ * Associate the event queue or completion queue to the epollfd wait mechanism.
  */
 static int libfab_waitfd_bind(struct fid* fid, struct m0_fab__tm *tm, void *ctx)
 {
-	struct epoll_event ev;
-	int                fd;
-	int                rc;
+	struct m0_fab__tm_fids *tmfid = &tm->ftm_fids;
+	struct m0_fab__ev_ctx  *ptr = ctx;
+	struct epoll_event      ev;
+	int                     fd;
+	int                     rc;
 
 	rc = fi_control(fid, FI_GETWAIT, &fd);
 	if (rc != 0)
 		return M0_ERR(rc);
 
 	ev.events = EPOLLIN;
-	ev.data.ptr = ctx;
+	ev.data.ptr = ptr;
+	M0_LOG(M0_DEBUG, "ADD_TO_EPOLL %s=%p fd=%d ctx=%p tm=%p pos=%d",
+	       ptr->evctx_dbg, fid, fd, ctx, tm, (int)tmfid->ftf_cnt);
 	rc = epoll_ctl(tm->ftm_epfd, EPOLL_CTL_ADD, fd, &ev);
 
+	if (rc == 0) {
+		if (tmfid->ftf_cnt >= (tmfid->ftf_arr_size - 1)) {
+			rc = libfab_fid_array_grow(tmfid,
+						   FAB_TM_FID_MALLOC_STEP);
+			if (rc != 0)
+				return M0_ERR(rc);
+		}
+		tmfid->ftf_head[tmfid->ftf_cnt] = fid;
+		tmfid->ftf_ctx[tmfid->ftf_cnt]  = ptr;
+		ptr->evctx_pos = tmfid->ftf_cnt;
+		tmfid->ftf_cnt++;
+		M0_ASSERT(tmfid->ftf_cnt < tmfid->ftf_arr_size);
+	}
+
+	return M0_RC(rc);
+}
+
+/**
+ * Remove the file descriptor of the event queue or completion queue from the
+ * interest list of the epoll fd.
+ */
+static int libfab_waitfd_unbind(struct fid* fid, struct m0_fab__tm *tm,
+				void *ctx)
+{
+	struct m0_fab__tm_fids *tmfid = &tm->ftm_fids;
+	struct m0_fab__ev_ctx  *ptr = ctx;
+	struct epoll_event      ev = {};
+	int                     fd;
+	int                     rc;
+	int                     i;
+
+	rc = fi_control(fid, FI_GETWAIT, &fd);
+	if (rc != 0)
+		return M0_ERR(rc);
+
+	rc = epoll_ctl(tm->ftm_epfd, EPOLL_CTL_DEL, fd, &ev);
+	if (rc == 0) {
+		M0_LOG(M0_DEBUG, "DEL_FROM_EPOLL %s fid=%p fd=%d tm=%p pos=%d",
+		       ptr->evctx_dbg, fid, fd, tm, ptr->evctx_pos);
+		for (i = ptr->evctx_pos; i < tmfid->ftf_cnt - 1; i++) {
+			tmfid->ftf_head[i] = tmfid->ftf_head[i + 1];
+			tmfid->ftf_ctx[i]  = tmfid->ftf_ctx[i + 1];
+			tmfid->ftf_ctx[i]->evctx_pos--;
+		}
+		--tmfid->ftf_cnt;
+		tmfid->ftf_head[tmfid->ftf_cnt] = 0;
+		tmfid->ftf_ctx[tmfid->ftf_cnt]  = 0;
+		ptr->evctx_pos = 0;
```

Comment:
Ideally should be set to invalid value e.g. -1, so that for loop will exit for invalid value 
for (i = ptr->evctx_pos; i < tmfid->ftf_cnt - 1; i++) {

Reply:
This will be taken up in the future task.

Reply:
okay.. lets do this as part of of address format